### PR TITLE
Enable reporting of TabletManagementIterator errors

### LINKED ISSUE #3469: Handle errors from plugins in tablet management iterator
In #3460 the TabletManagementIterator stated calling compaction related plugins to determine it a tablet needed compaction.  If these plugins throw an exception it could interfere with other manager functions like tablet assignment.  Need to handle this situation in such a way that other manager functions can continue and the minimum relevant information needed for debugging is logged.  Since the plugins are callled for each tablet, need to be careful about spamming the logs with lots or redundant information.

### PR DESCRIPTION
**Pull request by dlmarion**
Enable the TabletManagementIterator to report error's back to the TabletGroupWatcher when trying to determine actions that should be taken for an extent. Increment metric gauges when an error occurs.

Fixes #3469

### CODE REVIEW

**Review by keith-turner — COMMENTED** [review]
@dlmarion would it make sense to remove `ManagementAction.BAD_STATE` and roll it into this new more general error handling?  One nice thing about that is that it would cause the nice new metrics introduced in this PR to increment when that happens.

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java [hunk 1]
```diff
@@ -429,20 +429,33 @@ protected void consume() throws IOException {
...
-        // it was returned. Insert a K,V pair to represent the reasons. The client
-        // can pull this K,V pair from the results by looking at the colf.
-        TabletManagement.addActions(decodedRow, actions);
+      if (!actions.isEmpty() || error != null) {
+        if (error != null) {
+          // Insert the error into K,V pair representing
+          // the tablet metadata.
+          TabletManagement.addError(decodedRow, error);
+        }
+        if (!actions.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:452]: Looks like TabletGroupWatcher will not process the actions for a Tablet when there is an error, so I'm good with making this change.

File: core/src/main/java/org/apache/accumulo/core/metrics/MetricsProducer.java [hunk 2]
```diff
@@ -674,6 +695,11 @@ public interface MetricsProducer {
   String METRICS_GC_POST_OP_DURATION = METRICS_GC_PREFIX + "post.op.duration";
   String METRICS_GC_RUN_CYCLE = METRICS_GC_PREFIX + "run.cycle";
 
+  String METRICS_MANAGER_PREFIX = "accumulo.manager.";
+  String METRICS_MANAGER_ROOT_TGW_ERRORS = METRICS_MANAGER_PREFIX + "tgw.root.errors";
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/metrics/MetricsProducer.java]: Implemented suggestion in f6d4281

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java]: Implemented suggestion in f6d4281

File: core/src/main/java/org/apache/accumulo/core/manager/state/TabletManagement.java [hunk 3]
```diff
@@ -94,6 +115,10 @@ public TabletMetadata getTabletMetadata() {
     return tabletMetadata;
   }
 
+  public String getErrorMessage() {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply core/src/main/java/org/apache/accumulo/core/manager/state/TabletManagement.java:122]: Implemented suggestion in f6d4281

**Comment by dlmarion** [comment]
> @dlmarion would it make sense to remove `ManagementAction.BAD_STATE` and roll it into this new more general error handling? 

BAD_STATE is set in TabletManagementIterator when `tm.isFutureAndCurrentLocationSet()`. However, it's never checked in TabletGroupWatcher. Instead, TGW also checks `tm.isFutureAndCurrentLocationSet()` and throws `BadLocationStateException` in this case. When this is thrown, `repairMetadata` is called to fix it. I think we can get rid of BAD_STATE in the iterator entirely, as it's checked in TGW. I don't think we should roll it into the error handling.

**Comment by dlmarion** [comment]
Looking at the code further, I think we should leave as is. If we are not going to add an error for this condition, then we need to add an action so that the TabletMetadata gets returned.

**Review by keith-turner — APPROVED** [review]

**Comment by keith-turner** [comment]
> When this is thrown, repairMetadata is called to fix it. 

Ok, so need to do one or the other.  I had an item on the project board to look into the repairMetadata function and I just converted it to an issue #3888 and commented on it.

**Pull request merged**